README.md
<!-- 
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages). 

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages). 
-->

# Animated Page Reveal
**Animated Page Reveal** package lets you add an animated page reveal to your Flutter app that dispaly a set of pages with your desired content.


## Features
The AnimatedPageReveal widget is built to be a tool you can use to showcase a content in your Flutter app (ex: onBoarding screen). By default the position of content is set to `Center` and by using the `ViewPageModel`, you can add as many pages as you like with your desired content. 

* *Animated Pages*

By making a drag on the screen (leftToRight or rightToLeft) one page fades out and another comes in with a cool animation. The bottom icons indicate the position of the active page as well as both the seen & unseen pages remaining for the user to interact with.

*The package will handle the animation by itself.*

![animatedPageReveal](https://user-images.githubusercontent.com/68671238/135723534-5301fe7e-e2f7-4896-a94d-c31b84c2b8b0.png)
 
<hr>

## Getting started

1. Add the latest version of package to your `pubspec.yaml` (and run `dart pub get`):
```dart
dependencies:
  animated_page_reveal: ^1.0.0
```
2. Import the package and use it in your Flutter App.
```dart
import 'package:animated_page_reveal/animated_page_reveal.dart';
```
<hr>

## Usage

There are a number of properties that you can modify:

* title (Title text of the page)
* description (Body text of the page)
* color (Color of the page)
* imageAssetPath (Image of the page)
* iconAssetPath (Page bottom icon indicator)
* titleStyle (Title text Style)
* descriptionStyle (Body text style)

**Example Usage (Complete with all params):**
<table>
 <tr>
 <td>
      
```dart
class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: MediaQuery.of(context).size.width,
      height: MediaQuery.of(context).size.height,
      child: AnimatedPageReveal(children: [
        PageViewModel(
          title: 'Choose your place',
          description:
              'Pick the right destination according to the season because it is a key factor for a successful trip',
          color: const Color(0xff195932),
          imageAssetPath: 'assets/images/trip.png',
          iconAssetPath: 'assets/images/placeicon.png',
          titleStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 18,
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          descriptionStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 14,
            fontWeight: FontWeight.normal,
            color: Colors.white,
          ),
        ),
        PageViewModel(
          title: 'Book your flight',
          description:
              'Found a flight that matches your destination and schedule? \nBook it just a few taps',
          color: const Color(0xff19594E),
          imageAssetPath: 'assets/images/flight.png',
          iconAssetPath: 'assets/images/planicon.png',
          titleStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 18,
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          descriptionStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 14,
            fontWeight: FontWeight.normal,
            color: Colors.white,
          ),
        ),
        PageViewModel(
          title: 'Explore the world',
          description:
              'Easily discover new places and share them with your friends. \nMaybe you will plan together your next trip?',
          color: const Color(0xff193A59),
          imageAssetPath: 'assets/images/explore.png',
          iconAssetPath: 'assets/images/searchicon.png',
          titleStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 18,
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          descriptionStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 14,
            fontWeight: FontWeight.normal,
            color: Colors.white,
          ),
        ),
      ]),
    );
  }
}
```
   </td>
   <td>

   </td>
  </tr>
  </table>
<hr>

     Here's what it looks like:
         
https://user-images.githubusercontent.com/68671238/135723767-a464d5e3-2a88-4e10-b64d-af652be273ff.mp4



## Next Goals
We are working on some improvements including:

- [ ] Change color of the bottom circle icon indicator.
- [ ] Add a button to exit to another screen.

## Issues & Feedback
Please file an [issue!](https://github.com/aliMissaoui/Flutter-Package-Animated-Page-Reveal/issues) to send feedback or report a bug. Thank you!



CHANGELOG.md
## 1.0.0

* Animated page reveal to display content.


lib/src/pages.dart
import 'package:animated_page_reveal/src/page_view_model.dart';
import 'package:flutter/material.dart';

// Page content logic
class PageContent extends StatelessWidget {
  const PageContent(
      {Key? key, required this.viewModel, this.percentVisible = 1.0})
      : super(key: key);

  final PageViewModel viewModel;
  final double percentVisible;

  @override
  Widget build(BuildContext context) {
    return Container(
      // Give the page as much width as possible
      width: double.infinity,
      // Set the page color
      color: viewModel.color,
      // Wrap the content with opacity widget to control its visibility
      child: Opacity(
        opacity: percentVisible,
        child: Column(
          // Make the conetnt in center
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            // Transform widget for page image
            Transform(
              transform: Matrix4.translationValues(
                  0.0, 50.0 * (1 - percentVisible), 0.0),
              child: Padding(
                padding: const EdgeInsets.only(bottom: 25.0),
                child: Image.asset(
                  viewModel.imageAssetPath,
                  width: 200.0,
                  height: 200.0,
                ),
              ),
            ),
            // Transform widget for page title
            Transform(
              transform: Matrix4.translationValues(
                  0.0, 30.0 * (1 - percentVisible), 0.0),
              child: Padding(
                padding: const EdgeInsets.only(bottom: 10.0, top: 10.0),
                child: Text(
                  viewModel.title,
                  style: viewModel.titleStyle,
                ),
              ),
            ),
            // Transform widget for page description
            Transform(
              transform: Matrix4.translationValues(
                  0.0, 30.0 * (1 - percentVisible), 0.0),
              child: Padding(
                padding: const EdgeInsets.only(
                    bottom: 75.0, right: 20.0, left: 20.0),
                child: Text(
                  viewModel.description,
                  style: viewModel.descriptionStyle,
                  textAlign: TextAlign.center,
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}


lib/src/page_reveal.dart
import 'dart:math';

import 'package:flutter/material.dart';

// Page reveal logic
class PageReveal extends StatelessWidget {
  const PageReveal({Key? key, required this.revealPercent, required this.child})
      : super(key: key);

  final double revealPercent;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    return ClipOval(
      clipper: CircleRevealClipper(revealPercent),
      child: child,
    );
  }
}

class CircleRevealClipper extends CustomClipper<Rect> {
  final double revealPercent;

  CircleRevealClipper(this.revealPercent);

  @override
  Rect getClip(Size size) {
    final epicenter = Offset(size.width / 2, size.height * 0.9);

    // Calculate the distance from the epicenter to the left top corner to make sure we fill the screen.
    double theta = atan(epicenter.dy / epicenter.dx);

    final distanceToCorner = epicenter.dy / sin(theta);
    final radius = distanceToCorner * revealPercent;
    final diameter = 2 * radius;

    return Rect.fromLTWH(
        epicenter.dx - radius, epicenter.dy - radius, diameter, diameter);
  }

  @override
  bool shouldReclip(covariant CustomClipper<Rect> oldClipper) {
    return true;
  }
}


lib/src/page_dragger.dart
import 'dart:async';
import 'dart:ui';

import 'package:animated_page_reveal/src/pager_indicator.dart';
import 'package:flutter/material.dart';

// Page dragger logic
class PageDragger extends StatefulWidget {
  final bool canDragLeftToRight;
  final bool canDragRightToLeft;
  final StreamController<SlideUpdate> slideUpdateStream;
  const PageDragger(
      {Key? key,
      required this.slideUpdateStream,
      required this.canDragLeftToRight,
      required this.canDragRightToLeft})
      : super(key: key);

  @override
  _PageDraggerState createState() => _PageDraggerState();
}

class _PageDraggerState extends State<PageDragger> {
  static const fullTransitionPX = 300.0;
  Offset? dragStart;
  SlideDirection? slideDirection;
  double slidePercent = 0.0;

  onDragStart(DragStartDetails details) {
    dragStart = details.globalPosition;
  }

  onDragUpdate(DragUpdateDetails details) {
    if (dragStart != null) {
      final newPosition = details.globalPosition;
      final dx = dragStart!.dx - newPosition.dx;

      if (dx > 0.0 && widget.canDragRightToLeft) {
        slideDirection = SlideDirection.rightToLeft;
      } else if (dx < 0.0 && widget.canDragLeftToRight) {
        slideDirection = SlideDirection.leftToRight;
      } else {
        slideDirection = SlideDirection.none;
      }

      if (slideDirection != SlideDirection.none) {
        slidePercent = (dx / fullTransitionPX).abs().clamp(0.0, 1.0);
      } else {
        slidePercent = 0.0;
      }

      widget.slideUpdateStream.add(
        SlideUpdate(
          slideDirection!,
          slidePercent,
          UpdateType.dragging,
        ),
      );
    }
  }

  onDragEnd(DragEndDetails details) {
    widget.slideUpdateStream.add(
      SlideUpdate(
        SlideDirection.none,
        0.0,
        UpdateType.doneDragging,
      ),
    );
    dragStart = null;
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onHorizontalDragStart: onDragStart,
      onHorizontalDragUpdate: onDragUpdate,
      onHorizontalDragEnd: onDragEnd,
    );
  }
}

class AnimatedPageDragger {
  static const percentPerMilliSecond = 0.005;

  final SlideDirection slideDirection;
  final TransitionGoal transitionGoal;

  AnimationController? completionAnimationController;
  AnimatedPageDragger(this.slideDirection, this.transitionGoal, slidePercent,
      StreamController<SlideUpdate> slideUpdateStream, TickerProvider vsync) {
    final startlidePercent = slidePercent;
    double endSlidePercent;
    Duration duration;

    if (transitionGoal == TransitionGoal.open) {
      endSlidePercent = 1.0;
      final slideRemaining = 1.0 - slidePercent;
      duration = Duration(
          milliseconds: (slideRemaining / percentPerMilliSecond).round());
    } else {
      endSlidePercent = 0.0;
      duration = Duration(
          milliseconds: (slidePercent / percentPerMilliSecond).round());
    }

    completionAnimationController = AnimationController(
        duration: duration, vsync: vsync)
      ..addListener(() {
        slidePercent = lerpDouble(
          startlidePercent,
          endSlidePercent,
          completionAnimationController!.value,
        );
        slideUpdateStream.add(
            SlideUpdate(slideDirection, slidePercent, UpdateType.animating));
      })
      ..addStatusListener((AnimationStatus status) {
        if (status == AnimationStatus.completed) {
          slideUpdateStream.add(SlideUpdate(
              slideDirection, endSlidePercent, UpdateType.doneAnimating));
        }
      });
  }

  run() {
    completionAnimationController!.forward(from: 0.0);
  }

  dispose() {
    completionAnimationController!.dispose();
  }
}

enum TransitionGoal {
  open,
  close,
}
enum UpdateType {
  dragging,
  doneDragging,
  animating,
  doneAnimating,
}

class SlideUpdate {
  final UpdateType updateType;
  final SlideDirection direction;
  final double slidePercent;

  SlideUpdate(this.direction, this.slidePercent, this.updateType);
}


lib/src/page_view_model.dart
import 'package:flutter/material.dart';

/// Provides data for the pages.
class PageViewModel {
  /// Title of the page content.
  final String title;

  /// Description for the page content.
  final String description;

  /// Text Style of the page title.
  final TextStyle? titleStyle;

  /// Text Style of the page description.
  final TextStyle? descriptionStyle;

  /// Page color.
  final Color color;

  /// Main image of the page.
  final String imageAssetPath;

  /// bottom indicator image of the page.
  final String iconAssetPath;

  /// Provides data for the pages.
  PageViewModel({
    required this.title,
    required this.description,
    this.titleStyle,
    this.descriptionStyle,
    required this.color,
    required this.imageAssetPath,
    required this.iconAssetPath,
  });
}


lib/src/pager_indicator.dart
import 'dart:ui';

import 'package:animated_page_reveal/src/page_view_model.dart';
import 'package:flutter/material.dart';

// Page indicator logic
class PagerIndicator extends StatelessWidget {
  const PagerIndicator({Key? key, required this.viewModel}) : super(key: key);

  final PagerIndicatorViewModel viewModel;

  @override
  Widget build(BuildContext context) {
    List<PageBubble> bubbles = [];

    for (var i = 0; i < viewModel.pages.length; ++i) {
      final page = viewModel.pages[i];

      double percentActive;

      if (i == viewModel.activeIndex) {
        percentActive = 1.0 - viewModel.slidePercent;
      } else if (i == viewModel.activeIndex - 1 &&
          viewModel.slideDirection == SlideDirection.leftToRight) {
        percentActive = viewModel.slidePercent;
      } else if (i == viewModel.activeIndex + 1 &&
          viewModel.slideDirection == SlideDirection.rightToLeft) {
        percentActive = viewModel.slidePercent;
      } else {
        percentActive = 0.0;
      }

      bool isHollow = i > viewModel.activeIndex ||
          (i == viewModel.activeIndex &&
              viewModel.slideDirection == SlideDirection.leftToRight);

      bubbles.add(
        PageBubble(
          viewModel: PageBubbleViewModel(
            page.iconAssetPath,
            page.color,
            isHollow,
            percentActive,
          ),
        ),
      );
    }

    const bubbleWidth = 55.0;
    final baseTranslation =
        ((viewModel.pages.length * bubbleWidth) / 2) - (bubbleWidth / 2);
    var translation = baseTranslation - (viewModel.activeIndex * bubbleWidth);

    if (viewModel.slideDirection == SlideDirection.leftToRight) {
      translation += bubbleWidth * viewModel.slidePercent;
    } else if (viewModel.slideDirection == SlideDirection.rightToLeft) {
      translation -= bubbleWidth * viewModel.slidePercent;
    }
    return Column(
      children: [
        Expanded(child: Container()),
        Transform(
          transform: Matrix4.translationValues(translation, 0.0, 0.0),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: bubbles,
          ),
        ),
      ],
    );
  }
}

// Slide directions
enum SlideDirection {
  leftToRight,
  rightToLeft,
  none,
}

// View Model for page indicator
class PagerIndicatorViewModel {
  final List<PageViewModel> pages;
  final int activeIndex;
  final SlideDirection slideDirection;
  final double slidePercent;

  PagerIndicatorViewModel(
      this.pages, this.activeIndex, this.slideDirection, this.slidePercent);
}

// Page bubble logic
class PageBubble extends StatelessWidget {
  const PageBubble({Key? key, required this.viewModel}) : super(key: key);

  final PageBubbleViewModel viewModel;

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 55.0,
      height: 65.0,
      child: Center(
        child: Container(
          width: lerpDouble(20.0, 45.0, viewModel.activePercent),
          height: lerpDouble(20.0, 45.0, viewModel.activePercent),
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: viewModel.isHollow
                ? const Color(0x88ffffff)
                    .withAlpha((0x88 * viewModel.activePercent).round())
                : const Color(0x88ffffff),
            border: Border.all(
              color: viewModel.isHollow
                  ? const Color(0x88ffffff).withAlpha(
                      (0x88 * (1.0 - viewModel.activePercent)).round())
                  : Colors.transparent,
              width: 3.0,
            ),
          ),
          child: Opacity(
            opacity: viewModel.activePercent,
            child: Image.asset(
              viewModel.iconAssetPath,
              color: viewModel.color,
            ),
          ),
        ),
      ),
    );
  }
}

// View Model for page bubble
class PageBubbleViewModel {
  final String iconAssetPath;
  final Color color;
  final bool isHollow;
  final double activePercent;

  PageBubbleViewModel(
      this.iconAssetPath, this.color, this.isHollow, this.activePercent);
}


lib/src/animated_page_reveal_content.dart
import 'dart:async';

import 'package:animated_page_reveal/src/page_dragger.dart';
import 'package:animated_page_reveal/src/page_reveal.dart';
import 'package:animated_page_reveal/src/page_view_model.dart';
import 'package:animated_page_reveal/src/pager_indicator.dart';
import 'package:animated_page_reveal/src/pages.dart';
import 'package:flutter/material.dart';

class AnimatedPageReveal extends StatefulWidget {
  /// Children pages.
  final List<PageViewModel> children;

  const AnimatedPageReveal({Key? key, required this.children})
      : super(key: key);

  @override
  _AnimatedPageRevealState createState() => _AnimatedPageRevealState();
}

class _AnimatedPageRevealState extends State<AnimatedPageReveal>
    with TickerProviderStateMixin {
  StreamController<SlideUpdate> slideUpdateStream;
  AnimatedPageDragger? animatedPageDragger;

  int activeIndex = 0;
  int nextPageIndex = 0;
  SlideDirection slideDirection = SlideDirection.none;
  double slidePercent = 0.0;

  _AnimatedPageRevealState()
      : slideUpdateStream = StreamController<SlideUpdate>() {
    slideUpdateStream.stream.listen((SlideUpdate event) {
      setState(() {
        if (event.updateType == UpdateType.dragging) {
          slideDirection = event.direction;
          slidePercent = event.slidePercent;

          if (slideDirection == SlideDirection.leftToRight) {
            nextPageIndex = activeIndex - 1;
          } else if (slideDirection == SlideDirection.rightToLeft) {
            nextPageIndex = activeIndex + 1;
          } else {
            nextPageIndex = activeIndex;
          }
        } else if (event.updateType == UpdateType.doneDragging) {
          if (slidePercent > 0.5) {
            animatedPageDragger = AnimatedPageDragger(slideDirection,
                TransitionGoal.open, slidePercent, slideUpdateStream, this);
          } else {
            animatedPageDragger = AnimatedPageDragger(slideDirection,
                TransitionGoal.close, slidePercent, slideUpdateStream, this);

            nextPageIndex = activeIndex;
          }
          animatedPageDragger!.run();
        } else if (event.updateType == UpdateType.animating) {
          slideDirection = event.direction;
          slidePercent = event.slidePercent;
        } else if (event.updateType == UpdateType.doneAnimating) {
          activeIndex = nextPageIndex;
          slideDirection = SlideDirection.none;
          slidePercent = 0.0;

          animatedPageDragger!.dispose();
        }
      });
    });
  }
  @override
  Widget build(BuildContext context) {
    setContent(
        Color color,
        String title,
        String description,
        String imageAssetPath,
        String iconAssetPath,
        TextStyle? titleStyle,
        TextStyle? descriptionStyle) {
      return PageViewModel(
          color: color,
          title: title,
          titleStyle: titleStyle,
          description: description,
          descriptionStyle: descriptionStyle,
          iconAssetPath: iconAssetPath,
          imageAssetPath: imageAssetPath);
    }

    List<PageViewModel> _getChildrenList() {
      return widget.children.map((PageViewModel child) {
        return setContent(
            child.color,
            child.title,
            child.description,
            child.imageAssetPath,
            child.iconAssetPath,
            child.titleStyle,
            child.descriptionStyle);
      })
          // .toList()
          // .reversed
          .toList();
    }

    var pagesChildren = _getChildrenList();

    List<PageViewModel> pages = List.from(pagesChildren);
    return Scaffold(
      body: Stack(
        children: [
          PageContent(
            viewModel: pages[activeIndex],
            percentVisible: 1.0,
          ),
          PageReveal(
            revealPercent: slidePercent,
            child: PageContent(
              viewModel: pages[nextPageIndex],
              percentVisible: slidePercent,
            ),
          ),
          PagerIndicator(
            viewModel: PagerIndicatorViewModel(
                pages, activeIndex, slideDirection, slidePercent),
          ),
          PageDragger(
            slideUpdateStream: slideUpdateStream,
            canDragLeftToRight: activeIndex > 0,
            canDragRightToLeft: activeIndex < pages.length - 1,
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:animated_page_reveal/animated_page_reveal.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: MediaQuery.of(context).size.width,
      height: MediaQuery.of(context).size.height,
      child: AnimatedPageReveal(children: [
        PageViewModel(
          title: 'Choose your place',
          description:
              'Pick the right destination according to the season because it is a key factor for a successful trip',
          color: const Color(0xff195932),
          imageAssetPath: 'assets/images/trip.png',
          iconAssetPath: 'assets/images/placeicon.png',
          titleStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 18,
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          descriptionStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 14,
            fontWeight: FontWeight.normal,
            color: Colors.white,
          ),
        ),
        PageViewModel(
          title: 'Book your flight',
          description:
              'Found a flight that matches your destination and schedule? \nBook it just a few taps',
          color: const Color(0xff19594E),
          imageAssetPath: 'assets/images/flight.png',
          iconAssetPath: 'assets/images/planicon.png',
          titleStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 18,
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          descriptionStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 14,
            fontWeight: FontWeight.normal,
            color: Colors.white,
          ),
        ),
        PageViewModel(
          title: 'Explore the world',
          description:
              'Easily discover new places and share them with your friends. \nMaybe you will plan together your next trip?',
          color: const Color(0xff193A59),
          imageAssetPath: 'assets/images/explore.png',
          iconAssetPath: 'assets/images/searchicon.png',
          titleStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 18,
            fontWeight: FontWeight.bold,
            color: Colors.white,
          ),
          descriptionStyle: const TextStyle(
            fontFamily: 'Montserrat',
            fontSize: 14,
            fontWeight: FontWeight.normal,
            color: Colors.white,
          ),
        ),
      ]),
    );
  }
}
